Add APM support for Subscriptions and tokens with split UPE with deferred intent

james-allan opened the pull request:
Fixes #2872 

## Changes proposed in this Pull Request:

This PR adds/fixes the following:

1. Subscriptions via iDEAL, Bancontact and SEPA Debit.
2. Saved tokens via iDEAL, Bancontact and SEPA. 
   - Using saved tokens via block checkout.
3. Passing off-session args to the Stripe Payment Elements so the terms are shown to customers on the block checkout.
3. Refunds via payment methods that support it. 

## Testing instructions

### Subscriptions

1. Enable the iDEAL, SEPA, Bancontact payment methods.
2. Enable the Woo Subscriptions plugin. 
3. Create a subscription product if you don't have one. 
4. Add a subscription product to the cart. 
5. Go to the block and/or block checkout. 
   - On `add/deferred-intent` only the card payment element will be shown.
   - On this branch, SEPA, Bancontact and iDEAL should all be shown.

| `add/deferred-intent` | This branch |
|--------|--------|
| <img width="515" alt="Screenshot 2024-02-07 at 5 35 39 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/eb1fe8f9-2035-4d30-8436-8762877a1b20"> | <img width="518" alt="Screenshot 2024-02-07 at 5 36 05 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/eef7dc68-43ad-4e7e-bb6f-72a37ac28f29"> |

6. Test all these additional payment methods with subscriptions in your cart.
7. View the subscription and you should see that it has a payment method via SEPA. 

<p align="center">
<img width="460" alt="Screenshot 2024-02-07 at 5 38 18 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/c6c3bac5-53b3-42ab-9c95-896235b820b2">
</p>

> [!important]
> Keep in mind that no matter which payment method you use (SEPA, Bancontact or iDEAL) they all resolve to SEPA payment tokens as described in Stripe docs [here](https://stripe.com/docs/payments/bancontact), under "**Recurring payments**".

8. Test automatic, and early renewals. 
   - Note that because it's a SEPA token, final payment is processed via the webhook.

### Tokens

After you've purchased a subscription using any of the APMs mentioned above, you should now have a SEPA token. These steps will walk through the process of using a SEPA token and saving a new one. 

1. Add a simple product to your cart and on the checkout select Bancontact, SEPA, or iDEAL.
2. Choose to save the payment method. 
3. Complete checkout. 
4. Add another product to the cart and process the payment using the saved token. 
   - Again, note that because it's a SEPA token, final payment is processed via the webhook.


#### Deleting tokens 

1. Go into your database and the `wp_woocommerce_payment_tokens` table. 
2. Delete all your Stripe related tokens.  
3. Refresh the checkout page and notice that all the tokens you deleted that still exist in your Stripe account have been recreated. 

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

james-allan committed:
Add support flags to Stripe UPE methods

james-allan committed:
Set the wc-stripe-is-deferred-intent flag when processing payments via tokens on the block checkout

james-allan committed:
Add bancontact and ideal to the list of gateways that require a mandate

james-allan committed:
Set the subscriptions payment method ID when setting the payment token source

james-allan committed:
Fix saved tokens for APMs like iDEAL, Bancontact and SEPA

james-allan committed:
fix typo and fetch the payment method object

james-allan committed:
Make sure to pass offsession flag to block payment elements if the cart contains a subs
cription or saved tokens are enabled

james-allan committed:
Set the payment method ID on the subscription when the payment method is updated

james-allan commented:
I'm placing this one on-hold for a bit. I think there's an issue in the way I'm setting the subscription's payment method ID to `stripe` even if you use SEPA. I think the right way is to set the subscriptions payment method to the individual UPE method and then make the payment methods that support subscriptions call `$this->maybe_init_subscriptions();` to hook up all the relevant subscription hooks.

james-allan committed:
Set the order and subscription payment method to the upe gateway's ID

james-allan commented on client/blocks/upe/upe-deferred-intent-creation/payment-elements.js:
```diff
@@ -24,21 +24,25 @@ const PaymentElements = ( { api, ...props } ) => {
+		props.showSaveOption
+	) {
+		options.setupFutureUsage = 'off_session';
+	}
```
This change makes sure that when a subscription is in the cart, or if saving the method is allowed these terms are displayed to the user.

<img width="1047" alt="Screenshot 2024-02-09 at 2 28 40 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/51753b88-b322-4285-b8a5-f64fc3c350c9">

james-allan commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -364,6 +364,14 @@ public function add_payment_request_order_meta( PaymentContext $context, Payment
+				 */
+				if ( $is_stripe_payment_method && isset( $data['issavedtoken'] ) && $data['issavedtoken'] ) {
+					$context->set_payment_data( array_merge( $data, [ 'wc-stripe-is-deferred-intent' => true ] ) );
+				}
```
Using a saved payment token on the block checkout doesn't go via `PaymentProcessor` function and so this flag that we use to determine if we should process the request via the deferred intent flow isn't set. 

This change makes sure we set this flag early in the request to process the block checkout.

james-allan commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -893,12 +893,12 @@ private function get_payment_method_types_for_intent_creation( string $selected_
+		}
 
-		return $is_stripe_link_enabled || $is_sepa_debit_payment;
+		return 'card' === $selected_payment_type && in_array( 'link', $this->get_upe_gateway()->get_upe_enabled_payment_method_ids(), true );
```
Because Bancontact and iDEAL resolve to SEPA debit payment tokens, Stripe also requires that mandate data be set when attempting to use Bancontact and save it.

james-allan commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -911,11 +911,12 @@ public function is_mandate_data_required( $selected_payment_type ) {
 			'payment_method_types' => [ $payment_information['selected_payment_type'] ],
 			'customer'             => $payment_information['customer'],
 			'confirm'              => 'true',
+			'return_url'           => $payment_information['return_url'],
```
Now that Bancontact and iDEAL can be used for set up intents, we need to pass a return URL here.

james-allan committed:
Map payment method to the gateway ID should be recorded on the subscription

james-allan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -712,7 +714,7 @@ private function process_payment_with_deferred_intent( int $order_id ) {
 			// Handle saving the payment method in the store.
 			// It's already attached to the Stripe customer at this point.
-			if ( $payment_information['save_payment_method_to_store'] ) {
+			if ( $payment_information['save_payment_method_to_store'] && $upe_payment_method && $upe_payment_method->get_id() === $upe_payment_method->get_retrievable_type() ) {
```
This change can essentially be described as: only save the token if the payment method being used resolves to that same payment method. eg card = card 👍 bancontact = sepa ❌. 

When using a Bancontact to purchase a subscription we don't save the bancontact stripe payment method id (ie `pm_`) because it's not reusable. After the customer returns to the store after authorising the payment, stripe will create a `sepa_debit` token and we save it at that point. See `process_upe_redirect_payment()` and `process_order_for_confirmed_intent()`.

james-allan commented:
This is ready for review again. In the most recent set of changes I've changed it so when you purchase a subscription it will now set the subscription's payment gateway to the correct payment method ID. eg if you purchase it with a card, it will be set to `stripe`. If you purchase it with Bancontact, iDEAL or SEPA it will be set to `stripe_sepa_debit`.

This means the subscription will now list the right payment method type in the admin list table etc.

<img width="634" alt="Screenshot 2024-02-09 at 4 34 13 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/5dd99ba2-8a78-48ee-b60d-433ba44380d5">

mattallan reviewed (commented):
@james-allan I did some testing today and noticed a couple of things:

- :x: `pm_` token stored on the subscription is different than what gets saved in the wc_payment_tokens table
    - Token saved on the subscription:
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/3128698f-4dc2-41f7-87c6-ba3d7c64675e)
    - Token saved in the woo_payment_tokens table is different
    - From stripe, the one that is saved in the WC tokens table is the setup intent one, and the one saved on the subscription is the payment intent one:
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/cc312f01-ec3e-4237-ba61-48d5f3c06d9b)
- :x: Processing a renewal payment with an APM (ideal) results in the following fatal:
```
[12-Feb-2024 03:23:17 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method WC_Stripe_UPE_Payment_Method_Sepa::get_intent_from_order() in /Users/matt/local/woo/wp-content/plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php:818
Stack trace:
#0 /Users/matt/local/woo/wp-content/plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php(244): WC_Stripe_UPE_Payment_Method_Sepa->has_authentication_already_failed(Object(WC_Order))
#1 /Users/matt/local/woo/wp-content/plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php(189): WC_Stripe_UPE_Payment_Method_Sepa->process_subscription_payment('225.00', Object(WC_Order), true, false)
#2 /Users/matt/local/woo/wp-includes/class-wp-hook.php(324): WC_Stripe_UPE_Payment_Method_Sepa->scheduled_subscription_payment('225.00', Object(WC_Order))
#3 /Users/matt/local/woo/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#4 /Users/matt/local/woo/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#5 /Users/ in /Users/matt/local/woo/wp-content/plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php on line 818
```
- :x: On the Edit Subscription page, when you go to edit the payment method, there's two SEPA payment options. Inspecting the elements shows one is actually `'stripe'` but for some reason the title is being listed as SEPA? I couldn't figure this one out.

I've been digging into the second issue a bit and it brings up a bigger question (and potential issues). If the SEPA payment method is going to be using the subscriptions trait, should it extend a gateway class, or should we refactor the subscriptions trait so that it doesn't assume the trait will be used on a gateway class by using things like `WC_Stripe::get_instance()->get_main_stripe_gateway()->{gateway_method}` 🤔

james-allan committed:
Set payment method title and ID to the APM used

james-allan committed:
Make sure UPE methods can call methods on the main instance of the UPE gateway

james-allan committed:
Fix undefined $name error

james-allan commented:
@mattallan I've pushed https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/5f763772fd8c1e682346bb3e4b6c1512ac5365bd in line with the approach we discussed in a Slack huddle. I'm going to put it through some tests to make sure things are working as expected.

james-allan committed:
Merge branch 'add/deferred-intent' into fix/2872-dPE-APM-subscription-support

james-allan commented:
A couple of things I've noticed so far:

- [ ] With Multiple subscriptions in the cart iDEAL and Bancontact aren't being shown on the checkout. 
- [ ] Tokens are only backfilled on the classic checkout page. If you have tokens missing in your databse, loading the block checkout doesn't backfill them for some reason.
- [ ] Deleting token in Stripe directly doesn't delete it from token database in Woo. It's my understanding this was part of the `woocommerce_get_customer_upe_payment_tokens()` function's responsibility.

james-allan committed:
Reinstate the process payment function for upe methods

a-danae commented:
Thank you both for all the effort here!

I've created the following issues on top of this branch, hoping to distribute the load a bit:

- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2896 > To tackle "Tokens are only backfilled on the classic checkout page. If you have tokens missing in your database, loading the block checkout doesn't backfill them for some reason".
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2895 > To tackle "Deleting token in Stripe directly doesn't delete it from token database in Woo".
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2894 > Another side-effect to take care of.

Please feel free to mention if there are other behaviors that you think we could address in parallel.

james-allan committed:
Only attach hooks in the subscriptions trait once

mattallan commented:
I noticed my APM tokens aren't getting deleted from Stripe so I've opened a separate issue for this:
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2898

EDIT: opened another issue
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2899

james-allan commented:
Another issue: 

- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2900 - Changing the subscription's payment method to Bancontact or iDEAL isn't currently allowed. It theoretically should be possible.

mattallan commented:
@james-allan I did a bit more testing today and found some issues. Let me know if you'd rather move these out to separate issues so that we can divide and conquer 😄 

---

**APM testing (iDeal):**
- [x] Purchase simple
   - [x] Use new card
   - [x] Use saved card
- [x] Process renewal
   - [x] New card
   - [x] Saved card
   - [x] :x: If you have the SEPA payment method disabled in [WC > Settings > Payments > Stripe > Payment Methods](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/961e1024-7bbb-44c7-9ff5-5e1998af81e0). When you purchase a subscription with iDeal, the initial purchase is fine, but processing the renewal order fails with the following error:
```
The PaymentMethod provided (sepa_debit) is not allowed for this PaymentIntent. Please attach a PaymentMethod of one of the following types: card, giropay, eps, bancontact, ideal, p24. Alternatively, update the allowed payment_method_types for this PaymentIntent to include "sepa_debit"
```
- [x] Renew early
   - [x] Modal
   - [x] Checkout
- [x] Change payment method
   - :x: iDeal and Bancontact not showing up. This will be fixed by #2900
- [x] Purchase trial
   - [x] Use a saved card
   - [x] New card
- [x] Process switch
   - [x] No switch cost
   - [x] With prorated amount
- [x] Refund renewal order
   - :x: WC Core is calling `WC_Stripe_UPE_Payment_Method_Sepa::process_refund()` but it's not calling our UPE gateway refund function (😭), instead it's calling the `WC_Payment_Gateway::process_refund` which does nothing but `return false`. Looks like the `__call()` function isn't getting called in this case because the base class has defined this empty function.

**Standard Credit Card testing:**
- :x: Not sure if this is a known issue but while purchasing a subscription on the checkout block with a saved card the Subscription is active but requires manual renewal. Looking in the DB, the payment_method is being set to `stripe_card` instead of just `stripe` 😢. Should I move this to a separate issue?

james-allan committed:
Merge branch 'add/deferred-intent' into fix/2872-dPE-APM-subscription-support

james-allan commented:
> If you have the SEPA payment method disabled in [WC > Settings > Payments > Stripe > Payment Methods](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/961e1024-7bbb-44c7-9ff5-5e1998af81e0). When you purchase a subscription with iDeal, the initial purchase is fine, but processing the renewal order fails with the following error:

Good catch. I've had a look into this one, and the issue stems from this code in the `create_and_confirm_intent_for_off_session()` function. 

```php

if ( WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
	$payment_method_types = $this->get_upe_enabled_at_checkout_payment_method_ids();
} elseif ( isset( $prepared_source->source_object->type ) ) {
	$payment_method_types = [ $prepared_source->source_object->type ];
}
```

Given this function has **`off_session`** in the name, I'm not sure why it limits the allowed payment method types to those enabled at checkout. Its generally accepted that if a merchant disables a payment method, subscriptions already signed up with that payment method should still work, it just won't be available to new sign ups. ie all off session payments should be allowed right? 

With that in mind, I'm thinking of removing that first if statement entirely. Thoughts?

mattallan commented:
> With that in mind, I'm thinking of removing that first if statement entirely. Thoughts?

Yeah I agree. Looking at where `create_and_confirm_intent_for_off_session()` is called, it's only used when handling subscription and pre-order payments so it shouldn't be limited by enabled gateways only and should work for any stripe payment method imo.

james-allan committed:
Process UPE payment method refunds via the UPE gateway

james-allan committed:
Always use the source object's payment type when processing off session payments

james-allan commented:
> If you have the SEPA payment method disabled in [WC > Settings > Payments > Stripe > Payment Methods](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/961e1024-7bbb-44c7-9ff5-5e1998af81e0). When you purchase a subscription with iDeal, the initial purchase is fine, but processing the renewal order fails with the following error:

I've fixed that with https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/b99580b28551efcb05fae659236f863027e6eb8c.

> Refund renewal order ❌ 

I've fixed that in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/b13c1624f3e62a8f0ca758d477799f4d2bc67143.

> Not sure if this is a known issue but while purchasing a subscription on the checkout block with a saved card the Subscription is active but requires manual renewal. Looking in the DB, the payment_method is being set to stripe_card instead of just stripe 😢. Should I move this to a separate issue?

I still need to look into this one.

james-allan committed:
Use new helper function for determining the payment method id of a upe method object

james-allan commented:
> > Not sure if this is a known issue but while purchasing a subscription on the checkout block with a saved card the Subscription is active but requires manual renewal. Looking in the DB, the payment_method is being set to stripe_card instead of just stripe 😢. Should I move this to a separate issue?

> I still need to look into this one.

Ok. I've fixed that issue in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/d1f75410372b72eda0c6d1916356f9b60dad66b3 too. 😓

a-danae commented:
> Subscriptions via iDEAL, Bancontact and SEPA Debit.

Should we also include Sofort? It's reusable and it continues to be displayed when it was enabled before.

james-allan commented:
Here's the full set of tests I'm going to rerun through now that this PR seems a little more stable. I may add to them as I go. 

### Subscription sign up
- **_Simple Subscription Purchase_**
   - **Cards**
      - [x] New
      - [x] Saved token
      - [ ] ❌ New 3DS card - https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2905
      - [x] Saved 3DS card
   - **Bancontact**
      - [x] New
      - [x] Saved token
   - **iDEAL**
      - [x] New
      - [x] Saved token
   - **SEPA**
      - [x] New
      - [x] Saved token
   - **Sofort**
      - [x] New
      - [x] Saved token
- **_Free Trial Subscription Purchase_**
   - **Cards**
      - [x] New
      - [x] Saved token
      - [ ] New 3DS card
      - [x] Saved 3DS card
   - **Bancontact**
      - [x] New
      - [x] Saved token
   - **iDEAL**
      - [x] New
      - [x] Saved token
   - **SEPA**
      - [x] New
      - [x] Saved token
   - **Sofort**
      - [x] New
      - [x] Saved token
- **_Mixed carts (subscription + standard product)_**
   - **Cards**
      - [x] New
      - [x] Saved token
      - [ ] New 3DS card
      - [x] Saved 3DS card
   - **Bancontact**
      - [x] New
      - [x] Saved token
   - **iDEAL**
      - [x] New
      - [x] Saved token
   - **SEPA**
      - [x] New
      - [x] Saved token
   - **Sofort**
      - [x] New
      - [x] Saved token

### Renewals
- [x] Automatic renewals
- [x] Manual renewals
- [x] Early renewals (modal)
- [x] Early renewals (via cart)
- [x] Failed renewals (invalid token)
- [x] Failed renewal (insufficient funds token)

### Changing payment methods
- SEPA 
  - [x] Changing to new SEPA
  - [x] Change to existing SEPA token
- Card 
  - [ ] Change to new Card
  - [ ] Change to existing Card token
  - [ ] Change to new 3DS Card 
  - [ ] Change to existing 3DS Card token
 
Every other payment method will need to be part of: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2905
- Bancontact 
  - [ ] Change to new Bancontact
  - [ ] Change to existing Bancontact token
- iDEAL 
  - [ ] Change to new iDEAL
  - [ ] Change to existing iDEAL token
- Sofort 
  - [ ] Change to new iDEAL
  - [ ] Change to existing iDEAL token


### Tokens
- Deleting from My Account -> Payment Methods
- Adding tokens from My Account -> Payment Methods.

james-allan commented:
> Should we also include Sofort? It's reusable and it continues to be displayed when it was enabled before.

It's my understanding the Sofort has been discontinued by Stripe. There are PRs I've seen removing it from the plugin and in [Stripe's docs](https://docs.stripe.com/payments/sofort) they say that they are in the process of deprecating it but they do leave the door open to existing merchants.

So, if I've misunderstood and we do still support Sofort let me know. :)

mattallan commented:
> Deleting from My Account -> Payment Methods

Just making a note that Danae has fixed this in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2902 💯

a-danae committed:
Remove APM payment methods in Stripe when their tokens are deleted locally (#2902)

* Check whether the deleted token belongs to a reusable gateway

Before, we were only checking it belonged to the main gateway. Since Split PE, APMs have their own gateway so we needed to change this check

* Move the try/catch blocks up in the method

* Add Bancontact, Ideal, and Sofort to the static list of reusable gateways

a-danae commented:
> It's my understanding the Sofort has been discontinued by Stripe. There are PRs I've seen removing it from the plugin and in [Stripe's docs](https://docs.stripe.com/payments/sofort) they say that they are in the process of deprecating it but they do leave the door open to existing merchants.

Yup, that's what I've understood: It's been discontinued by Stripe but still functioning for existing merchants.

In the plugin, looks like we haven't removed it for merchants that still have it active ([ref in UPE](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/deferred-intent/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L136-L141), [ref in classic](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/deferred-intent/includes/class-wc-stripe-helper.php#L366-L373)).
That makes me think we should continue to make it functional, so it continues to work for merchants that have it active. Glad to learn if that's not the case, tho!

james-allan committed:
Add subscription support for Sofort

james-allan commented:
> That makes me think we should continue to make it functional

Ok, yeah that makes sense. I've updated the Sofort UPE method to now support subscriptions in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/d6690c95279036d15a5a7a82b2286fe403ae2232. I've also updated my list of tests above to also test Sofort. 

FWIW and incase this is helpful to anyone else, because my Stipe settings didn't show Sofort as an available payment method, I just forced it on via the option.

```php
$settings = get_option( 'woocommerce_stripe_settings' );

$settings['upe_checkout_experience_accepted_payments'][] = 'sofort';
update_option( 'woocommerce_stripe_settings', $settings );
```

mattallan reviewed (approved):
@james-allan I did a bunch more testing on this branch on the latest changes today and I would say this is ready to be merged!!! 👏 

I found one issue with not being able to use a saved credit card when paying for a failed subscription renewal. I was able to replicate this on the base branch `add/deferred-intent` so I'm going to create a separate issue to fix this.

Question:
 1. Should iDeal and Bancontact support multiple subscriptions or leave it out for now?

---

### iDeal
- [x] Purchase simple
    - [x] Use new card
    - [x] Use saved card
- [ ] Purchase multiple subscriptions
    - iDeal not supported ❓
- [x] Process renewal
    - [x] New card
    - [x] Saved card
    - [x] Confirm successful renewals with iDeal even with the SEPA payment method disabled 
- [x] Renew early
    - [x] Modal
    - [x] Checkout
- [ ] Change payment method
    - iDeal/Bancontact not listed. Will be fixed separately in #2900
- [x] Purchase trial
    - [x] Use a saved card
    - [x] New card
- [x] Process switch
    - [x] No switch cost
    - [x] With prorated amount
- [x] Refund renewal order

### Credit Card
- [x] Purchase simple
    - [x] Use new card
        - [x] Checkout shortcode
        - [x] Checkout block
    - [x] Use saved card
        - [x] Checkout shortcode
        - [x] Checkout block
- [x] Purchase multiple subscriptions
    - [x] Use new card
        - [x] Checkout shortcode
        - [x] Checkout block
    - [x] Use saved card
        - [x] Checkout block
        - [x] Checkout shortcode
- [x] Purchase mixed checkout
    - [x] Use new card
        - [x] Checkout shortcode
        - [x] Checkout block
    - [x] Use saved card
        - [x] Checkout block
        - [x] Checkout shortcode
- [x] Process renewals
    - [x] single purchase
    - [x] multiple subscriptions
    - [x] mixed checkout
- [x] Renew early
    - [x] Modal
    - [x] Checkout
- [x] Change payment method
- [x] Purchase trial
    - [x] Use new card
        - [x] Checkout shortcode
        - [x] Checkout block
    - [x] Use saved card
        - [x] Checkout block
        - [x] Checkout shortcode
- [x] Process switch
- [x] Refund renewal order
- [ ] Pay for a failed renewal
    - [x] New card
        - [x] Checkout shortcode
        - [x] Checkout block
    - [ ] Saved card
        - [x] Checkout shortcode
        - [ ] Checkout block :x:
            - Error found in stripe logs: No such payment_intent: 'seti_XXXXXX'
            - Can reproduce on add/deferred-intent so I'll create a separate issue for this

### SEPA Direct Debit
- [x] Purchase subscriptiom
    - [x] Use new card
        - [x] AT611904300234573201
        - [x] AT861904300235473202 (fails)
    - [x] Use saved card
- [x] Renew
- [x] Renew early
- [x] Change payment method
    - [x] change to different SEPA card
    - [x] change to a credit card
- [x] Purchase trial
- [x] Refund renewal order

### Bancontact
- [x] Purchase simple
    - [x] Use new card
    - [x] Use saved card
- [x] Renew
- [x] Renew early
- [x] Purchase trial
- [x] Refund renewal order

james-allan commented:
@mattallan I'm still going through your latest response and all the tests and questions you've raised. I was just in the middle of my tests too and I found an issue with using a new 3DS card on the block checkout (I haven't tested classic checkout). The purchase goes through but the payment method on the subscriptions says `N/A` and the customer ID is for some reason missing. 

| <img width="631" alt="Screenshot 2024-02-15 at 1 18 00 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/13e2fc48-ea5e-475b-b906-d8be95a3d4d8"> | <img width="306" alt="Screenshot 2024-02-15 at 1 22 15 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/ef3ddcd5-5255-42aa-ac00-72f5fa14a380"> |
|--------|--------|

Can you just check that you can replicate that too and I'll file an issue for it.

mattallan commented:
> Can you just check that you can replicate that too and I'll file an issue for it.

Yep I can reproduce it, but only on the Checkout block page and only when using a new card.

- 👍 Existing 3DS saved card on block checkout
- :x: New 3DS card on block checkout
- 👍 Existing 3DS saved card on shortcode checkout
- :+1: New 3DS card on shortcode checkout

Payment method on My Account > View subscription is N/A, but the payment method in the DB is still set to `stripe`.

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/32a95aac-e9e3-4c39-a0b1-43a404e5ed32)

james-allan commented:
Thanks for confirming. I can also replicate on `add/deferred-intent` so will file an issue. It appears that the issue is caused by the fact that the `maybe_process_upe_redirect()` is never ran when using the block checkout in this scenario and that means that the completed/authorized payment source isn't processed.

mattallan commented:
> [!NOTE]
> I've opened a new issue to address the problem I found to do with paying for a failed renewal using a saved card on the block checkout: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2904/

james-allan committed:
Add multiple subscription support flags to Bancontact, iDEAL, Sofort methods

james-allan commented:
> Should iDeal and Bancontact support multiple subscriptions or leave it out for now?

I've patched that in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2883/commits/e1e3926a932971d2bbb4f1950375843d1c329293. :)
